File: src/main/scala/com/stellmangreene/pbprdf/model/Ontology.scala
package com.stellmangreene.pbprdf.model

import org.eclipse.rdf4j.model.impl.SimpleValueFactory

/**
 * RDF types
 *
 * @author andrewstellman
 */
object Ontology {

  private val valueFactory = SimpleValueFactory.getInstance

  val NAMESPACE = "http://stellman-greene.com/pbprdf#"

  // Prefixes

  @OntologyPrefix(prefix = "players")
  val PLAYERS_PREFIX = "http://stellman-greene.com/pbprdf/players/"

  // Basic entity types

  @OntologyClass(label = "A game")
  val GAME = valueFactory.createIRI(NAMESPACE, "Game")

  @OntologyClass(label = "An event in a play-by-play")
  var EVENT = valueFactory.createIRI(NAMESPACE, "Event")

  @OntologyClass(label = "A play in a play-by-play")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event"))
  val PLAY = valueFactory.createIRI(NAMESPACE, "Play")

  @OntologyClass(label = "A team")
  val TEAM = valueFactory.createIRI(NAMESPACE, "Team")

  @OntologyClass(label = "A player")
  val PLAYER = valueFactory.createIRI(NAMESPACE, "Player")

  // Properties and types for each game

  @OntologyProperty(
    label = "The game time",
    domain = "http://stellman-greene.com/pbprdf#Game",
    range = "http://www.w3.org/2001/XMLSchema#dateTime")
  val GAME_TIME = valueFactory.createIRI(NAMESPACE, "gameTime")

  @OntologyProperty(
    label = "The game location",
    domain = "http://stellman-greene.com/pbprdf#Game",
    range = "http://www.w3.org/2001/XMLSchema#string")
  val GAME_LOCATION = valueFactory.createIRI(NAMESPACE, "gameLocation")

  @OntologyProperty(
    label = "The home team for a game",
    domain = "http://stellman-greene.com/pbprdf#Game",
    range = "http://stellman-greene.com/pbprdf#Team")
  @OntologyObjectProperty
  val HOME_TEAM = valueFactory.createIRI(NAMESPACE, "homeTeam")

  @OntologyProperty(
    label = "The away team for a game",
    domain = "http://stellman-greene.com/pbprdf#Game",
    range = "http://stellman-greene.com/pbprdf#Team")
  @OntologyObjectProperty
  val AWAY_TEAM = valueFactory.createIRI(NAMESPACE, "awayTeam")

  @OntologyProperty(
    label = "The home team roster for a game",
    domain = "http://stellman-greene.com/pbprdf#Game",
    range = "http://stellman-greene.com/pbprdf#Roster")
  @OntologyObjectProperty
  @OntologyComment(comment = "The home team roster is typically represented by a bnode in the Game entity")
  val HAS_HOME_TEAM_ROSTER = valueFactory.createIRI(NAMESPACE, "hasHomeTeamRoster")

  @OntologyProperty(
    label = "The away team roster for a game",
    domain = "http://stellman-greene.com/pbprdf#Game",
    range = "http://stellman-greene.com/pbprdf#Roster")
  @OntologyObjectProperty
  @OntologyComment(comment = "The away team roster is typically represented by a bnode in the Game entity")
  val HAS_AWAY_TEAM_ROSTER = valueFactory.createIRI(NAMESPACE, "hasAwayTeamRoster")

  @OntologyClass(label = "A game roster")
  val ROSTER = valueFactory.createIRI(NAMESPACE, "Roster")

  @OntologyProperty(
    label = "The team for a home or away roster",
    domain = "http://stellman-greene.com/pbprdf#Roster",
    range = "http://stellman-greene.com/pbprdf#Team")
  @OntologyObjectProperty
  val ROSTER_TEAM = valueFactory.createIRI(NAMESPACE, "rosterTeam")

  @OntologyProperty(
    label = "A player on a roster",
    domain = "http://stellman-greene.com/pbprdf#Roster",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val HAS_PLAYER = valueFactory.createIRI(NAMESPACE, "hasPlayer")

  // Types and properties for events in a play-by-play

  @OntologyProperty(
    label = "The game the event occurred in",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://stellman-greene.com/pbprdf#Game")
  val IN_GAME = valueFactory.createIRI(NAMESPACE, "inGame")

  @OntologyProperty(
    label = "The team that an event is for",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://stellman-greene.com/pbprdf#Team")
  @OntologyObjectProperty
  val FOR_TEAM = valueFactory.createIRI(NAMESPACE, "forTeam")

  @OntologyProperty(
    label = "The game time of an event (eg. 9:37)",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://www.w3.org/2001/XMLSchema#string")
  val TIME = valueFactory.createIRI(NAMESPACE, "time")

  @OntologyProperty(
    label = "The period the event occurred in",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://www.w3.org/2001/XMLSchema#int")
  @OntologyComment(comment = "Regulation periods are 1 through 4, overtime periods start at 5")
  val PERIOD = valueFactory.createIRI(NAMESPACE, "period")

  @OntologyProperty(
    label = "The number of seconds into a game that an event occurred",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://www.w3.org/2001/XMLSchema#int")
  val SECONDS_INTO_GAME = valueFactory.createIRI(NAMESPACE, "secondsIntoGame")

  @OntologyProperty(
    label = "The number of seconds left in the period when an event occurred",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://www.w3.org/2001/XMLSchema#int")
  val SECONDS_LEFT_IN_PERIOD = valueFactory.createIRI(NAMESPACE, "secondsLeftInPeriod")

  // Types and properties for end of period or game

  @OntologyClass(label = "End of period")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val END_OF_PERIOD = valueFactory.createIRI(NAMESPACE, "EndOfPeriod")

  @OntologyClass(label = "End of game")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val END_OF_GAME = valueFactory.createIRI(NAMESPACE, "EndOfGame")

  // Types and properties for timeouts

  @OntologyClass(label = "A timeout")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val TIMEOUT = valueFactory.createIRI(NAMESPACE, "Timeout")

  @OntologyProperty(
    label = "The duration of a timeout",
    domain = "http://stellman-greene.com/pbprdf#Timeout",
    range = "http://www.w3.org/2001/XMLSchema#string")
  @OntologyComment(comment = "A string description (eg. Full timeout, 20 Sec.")
  val TIMEOUT_DURATION = valueFactory.createIRI(NAMESPACE, "timeoutDuration")

  @OntologyProperty(
    label = "Determines if a timeout is an official timeout",
    domains = Array("http://stellman-greene.com/pbprdf#Timeout"),
    range = "http://www.w3.org/2001/XMLSchema#boolean")
  @OntologyComment(comment = "True for an official timeout")
  val IS_OFFICIAL = valueFactory.createIRI(NAMESPACE, "isOfficial")

  // Types and properties for jump balls

  @OntologyClass(label = "A jump ball")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val JUMP_BALL = valueFactory.createIRI(NAMESPACE, "JumpBall")

  @OntologyProperty(
    label = "The home player contesting the jump ball",
    domain = "http://stellman-greene.com/pbprdf#JumpBall",
    range = "http://stellman-greene.com/pbprdf#Player")
  val JUMP_BALL_HOME_PLAYER = valueFactory.createIRI(NAMESPACE, "jumpBallHomePlayer")

  @OntologyProperty(
    label = "The away player contesting the jump ball",
    domain = "http://stellman-greene.com/pbprdf#JumpBall",
    range = "http://stellman-greene.com/pbprdf#Player")
  val JUMP_BALL_AWAY_PLAYER = valueFactory.createIRI(NAMESPACE, "jumpBallAwayPlayer")

  @OntologyProperty(
    label = "The player who gained possession of the jump ball",
    domain = "http://stellman-greene.com/pbprdf#JumpBall",
    range = "http://stellman-greene.com/pbprdf#Player")
  val JUMP_BALL_GAINED_POSSESSION = valueFactory.createIRI(NAMESPACE, "jumpBallGainedPossession")

  // Rebounds

  @OntologyClass(label = "A rebound")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val REBOUND = valueFactory.createIRI(NAMESPACE, "Rebound")

  @OntologyProperty(
    label = "The player who rebounded the ball",
    domain = "http://stellman-greene.com/pbprdf#Rebound",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val REBOUNDED_BY = valueFactory.createIRI(NAMESPACE, "reboundedBy")

  @OntologyProperty(
    label = "Determines if a foul or rebound is offensive",
    domains = Array("http://stellman-greene.com/pbprdf#Foul", "http://stellman-greene.com/pbprdf#Rebound"),
    range = "http://www.w3.org/2001/XMLSchema#boolean")
  @OntologyComment(comment = "Only ever has the value xsd:true, no value exists if the play was defensive")
  val IS_OFFENSIVE = valueFactory.createIRI(NAMESPACE, "isOffensive")

  // Shots and blocks

  @OntologyClass(label = "A shot")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val SHOT = valueFactory.createIRI(NAMESPACE, "Shot")

  @OntologyClass(label = "A block")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play", "http://stellman-greene.com/pbprdf#Shot"))
  val BLOCK = valueFactory.createIRI(NAMESPACE, "Block") // pbprdf:Block rdfs:isSubclassOf pbprdf:Shot

  @OntologyProperty(
    label = "True if a shot was made",
    domain = "http://stellman-greene.com/pbprdf#Shot",
    range = "http://www.w3.org/2001/XMLSchema#boolean")
  @OntologyComment(comment = "Can have the value xsd:true for made shots or xsd:false for missed shots, will always be present for a shot")
  val SHOT_MADE = valueFactory.createIRI(NAMESPACE, "shotMade")

  @OntologyProperty(
    label = "The player that attempted the shot",
    domain = "http://stellman-greene.com/pbprdf#Shot",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val SHOT_BY = valueFactory.createIRI(NAMESPACE, "shotBy")

  @OntologyProperty(
    label = "The type of shot",
    domain = "http://stellman-greene.com/pbprdf#Shot",
    range = "http://www.w3.org/2001/XMLSchema#string")
  @OntologyComment(comment = "Will be a description such as layup or 21-foot jumper")
  val SHOT_TYPE = valueFactory.createIRI(NAMESPACE, "shotType")

  @OntologyProperty(
    label = "The player that assisted the shot",
    domain = "http://stellman-greene.com/pbprdf#Shot",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val SHOT_ASSISTED_BY = valueFactory.createIRI(NAMESPACE, "shotAssistedBy")

  @OntologyProperty(
    label = "The number of points scored by the shot",
    domain = "http://stellman-greene.com/pbprdf#Shot",
    range = "http://www.w3.org/2001/XMLSchema#int")
  @OntologyComment(comment = "Will only be present for made shots, will contain 1 for free throws, 2 for field goals, etc.")
  val SHOT_POINTS = valueFactory.createIRI(NAMESPACE, "shotPoints")

  @OntologyProperty(
    label = "The player that blocked the shot",
    domain = "http://stellman-greene.com/pbprdf#Block",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val SHOT_BLOCKED_BY = valueFactory.createIRI(NAMESPACE, "shotBlockedBy")

  // Fouls

  @OntologyClass(label = "A foul")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val FOUL = valueFactory.createIRI(NAMESPACE, "Foul")

  @OntologyProperty(
    label = "The player who committed the foul",
    domain = "http://stellman-greene.com/pbprdf#Foul",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val FOUL_COMMITTED_BY = valueFactory.createIRI(NAMESPACE, "foulCommittedBy")

  @OntologyProperty(
    label = "The player who drew the foul",
    domain = "http://stellman-greene.com/pbprdf#Foul",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val FOUL_DRAWN_BY = valueFactory.createIRI(NAMESPACE, "foulDrawnBy")

  @OntologyProperty(
    label = "Determines if a foul is a shooting foul",
    domain = "http://stellman-greene.com/pbprdf#Foul",
    range = "http://www.w3.org/2001/XMLSchema#boolean")
  @OntologyComment(comment = "Only ever has the value xsd:true, no value exists if the play was not a shooting foul")
  val IS_SHOOTING_FOUL = valueFactory.createIRI(NAMESPACE, "isShootingFoul")

  @OntologyProperty(
    label = "Determines if a foul is a loose ball foul",
    domain = "http://stellman-greene.com/pbprdf#Foul",
    range = "http://www.w3.org/2001/XMLSchema#boolean")
  @OntologyComment(comment = "Only ever has the value xsd:true, no value exists if the play was not a loose ball foul")
  val IS_LOOSE_BALL_FOUL = valueFactory.createIRI(NAMESPACE, "isLooseBallFoul")

  @OntologyProperty(
    label = "Determines if a foul is a charge",
    domain = "http://stellman-greene.com/pbprdf#Foul",
    range = "http://www.w3.org/2001/XMLSchema#boolean")
  @OntologyComment(comment = "Only ever has the value xsd:true, no value exists if the play was not a charge")
  val IS_CHARGE = valueFactory.createIRI(NAMESPACE, "isCharge")

  @OntologyProperty(
    label = "Determines if a foul is a personal blocking foul",
    domains = Array("http://stellman-greene.com/pbprdf#Foul"),
    range = "http://www.w3.org/2001/XMLSchema#boolean")
  @OntologyComment(comment = "Only ever has the value xsd:true, no value exists if the play was not a personal blocking foul")
  val IS_PERSONAL_BLOCKING_FOUL = valueFactory.createIRI(NAMESPACE, "isPersonalBlockingFoul")

  // Technicals

  @OntologyClass(label = "A technical foul")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val TECHNICAL_FOUL = valueFactory.createIRI(NAMESPACE, "TechnicalFoul")

  @OntologyProperty(
    label = "Determines if a technical foul is a three second violation",
    domain = "http://stellman-greene.com/pbprdf#TechnicalFoul",
    range = "http://www.w3.org/2001/XMLSchema#boolean")
  @OntologyComment(comment = "Only ever has the value xsd:true, no value exists if the play was not a three second violation")
  val IS_THREE_SECOND = valueFactory.createIRI(NAMESPACE, "isThreeSecond")

  @OntologyProperty(
    label = "Determines if a technical foul is a delay of game violation",
    domain = "http://stellman-greene.com/pbprdf#TechnicalFoul",
    range = "http://www.w3.org/2001/XMLSchema#boolean")
  @OntologyComment(comment = "Only ever has the value xsd:true, no value exists if the play was not a delay of game violation")
  val IS_DELAY_OF_GAME = valueFactory.createIRI(NAMESPACE, "isDelayOfGame")

  @OntologyProperty(
    label = "The technical foul number for a player",
    domain = "http://stellman-greene.com/pbprdf#TechnicalFoul",
    range = "http://www.w3.org/2001/XMLSchema#int")
  @OntologyComment(comment = "Will be 1 for first technical foul, 2 for 2nd")
  val TECHNICAL_FOUL_NUMBER = valueFactory.createIRI(NAMESPACE, "technicalFoulNumber")

  @OntologyClass(label = "A five-second violation")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val FIVE_SECOND_VIOLATION = valueFactory.createIRI(NAMESPACE, "FiveSecondViolation")

  // Ejection

  @OntologyClass(label = "An ejection")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val EJECTION = valueFactory.createIRI(NAMESPACE, "Ejection")

  @OntologyProperty(
    label = "The player that was ejected",
    domain = "http://stellman-greene.com/pbprdf#Ejection",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val PLAYER_EJECTED = valueFactory.createIRI(NAMESPACE, "playerEjected")

  // Players entering

  @OntologyClass(label = "A player entering the game")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val ENTERS = valueFactory.createIRI(NAMESPACE, "Enters")

  @OntologyProperty(
    label = "The player entering the game",
    domain = "http://stellman-greene.com/pbprdf#Enters",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val PLAYER_ENTERING = valueFactory.createIRI(NAMESPACE, "playerEntering")

  @OntologyProperty(
    label = "The player exiting the game",
    domain = "http://stellman-greene.com/pbprdf#Enters",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val PLAYER_EXITING = valueFactory.createIRI(NAMESPACE, "playerExiting")

  // Turnovers

  @OntologyClass(label = "A turnover")
  @OntologySubClassOf(subClassOf = Array("http://stellman-greene.com/pbprdf#Event", "http://stellman-greene.com/pbprdf#Play"))
  val TURNOVER = valueFactory.createIRI(NAMESPACE, "Turnover")

  @OntologyProperty(
    label = "The type of turnover",
    domain = "http://stellman-greene.com/pbprdf#Turnover",
    range = "http://www.w3.org/2001/XMLSchema#string")
  @OntologyComment(comment = "Contains text that describs the turnover (eg. traveling, kicked ball violation, etc.")
  val TURNOVER_TYPE = valueFactory.createIRI(NAMESPACE, "turnoverType")

  @OntologyProperty(
    label = "The player who stole the ball",
    domain = "http://stellman-greene.com/pbprdf#Turnover",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val STOLEN_BY = valueFactory.createIRI(NAMESPACE, "stolenBy")

  @OntologyProperty(
    label = "The player who turned the ball over",
    domain = "http://stellman-greene.com/pbprdf#Turnover",
    range = "http://stellman-greene.com/pbprdf#Player")
  @OntologyObjectProperty
  val TURNED_OVER_BY = valueFactory.createIRI(NAMESPACE, "turnedOverBy")

  // Previous and next events

  @OntologyProperty(
    label = "The event number",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://www.w3.org/2001/XMLSchema#int")
  val EVENT_NUMBER = valueFactory.createIRI(NAMESPACE, "eventNumber")

  @OntologyProperty(
    label = "The previous event in the game",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://stellman-greene.com/pbprdf#Event")
  @OntologyObjectProperty
  val PREVIOUS_EVENT = valueFactory.createIRI(NAMESPACE, "previousEvent")

  @OntologyProperty(
    label = "The next event in the game",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://stellman-greene.com/pbprdf#Event")
  @OntologyObjectProperty
  val NEXT_EVENT = valueFactory.createIRI(NAMESPACE, "nextEvent")

  @OntologyProperty(
    label = "The number of seconds since the previous event",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://www.w3.org/2001/XMLSchema#int")
  val SECONDS_SINCE_PREVIOUS_EVENT = valueFactory.createIRI(NAMESPACE, "secondsSincePreviousEvent")

  @OntologyProperty(
    label = "The number of seconds until the next event",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://www.w3.org/2001/XMLSchema#int")
  val SECONDS_UNTIL_NEXT_EVENT = valueFactory.createIRI(NAMESPACE, "secondsUntilNextEvent")

  // Scores

  @OntologyProperty(
    label = "Score for the home team",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://www.w3.org/2001/XMLSchema#int")
  val HOME_SCORE = valueFactory.createIRI(NAMESPACE, "homeScore")

  @OntologyProperty(
    label = "Score for the away team",
    domain = "http://stellman-greene.com/pbprdf#Event",
    range = "http://www.w3.org/2001/XMLSchema#int")
  val AWAY_SCORE = valueFactory.createIRI(NAMESPACE, "awayScore")

}

File: src/pbprdf/ontology.py
from __future__ import annotations

from dataclasses import dataclass

from rdflib import Graph, Literal, Namespace, URIRef
from rdflib.namespace import OWL, RDF, RDFS, XSD

PBPRDF = Namespace("http://stellman-greene.com/pbprdf#")
PBPRDF_ENTITY = Namespace("http://stellman-greene.com/pbprdf/")


@dataclass(frozen=True)
class PropertySpec:
    name: str
    domain: str
    range: str
    label: str
    object_property: bool = False
    subproperty_of: str | None = None


CLASS_NAMES = [
    "Game",
    "Event",
    "Play",
    "Team",
    "Player",
    "Roster",
    "Shot",
    "Block",
    "Foul",
    "TechnicalFoul",
    "Turnover",
    "Rebound",
    "JumpBall",
    "Timeout",
    "Enters",
    "Ejection",
    "EndOfPeriod",
    "EndOfGame",
    "FiveSecondViolation",
    "PlayType",
    "CourtCoordinate",
    "GameFormat",
    "WinProbabilitySnapshot",
    "Venue",
    "Official",
]


PROPERTY_SPECS = [
    # Game + roster
    PropertySpec("gameTime", "Game", str(XSD.dateTime), "gameTime"),
    PropertySpec("gameLocation", "Game", str(XSD.string), "gameLocation"),
    PropertySpec("homeTeam", "Game", "Team", "homeTeam", object_property=True),
    PropertySpec("awayTeam", "Game", "Team", "awayTeam", object_property=True),
    PropertySpec("hasHomeTeamRoster", "Game", "Roster", "hasHomeTeamRoster", object_property=True),
    PropertySpec("hasAwayTeamRoster", "Game", "Roster", "hasAwayTeamRoster", object_property=True),
    PropertySpec("rosterTeam", "Roster", "Team", "rosterTeam", object_property=True),
    PropertySpec("hasPlayer", "Roster", "Player", "hasPlayer", object_property=True),
    PropertySpec("espnEventId", "Game", str(XSD.string), "espnEventId"),
    PropertySpec("hasGameFormat", "Game", "GameFormat", "hasGameFormat", object_property=True),
    # Game format
    PropertySpec("leagueCode", "GameFormat", str(XSD.string), "leagueCode"),
    PropertySpec("regulationPeriodType", "GameFormat", str(XSD.string), "regulationPeriodType"),
    PropertySpec("regulationPeriodCount", "GameFormat", str(XSD.int), "regulationPeriodCount"),
    PropertySpec("regulationPeriodLengthMinutes", "GameFormat", str(XSD.int), "regulationPeriodLengthMinutes"),
    PropertySpec("overtimePeriodLengthMinutes", "GameFormat", str(XSD.int), "overtimePeriodLengthMinutes"),
    # Event core
    PropertySpec("inGame", "Event", "Game", "inGame", object_property=True),
    PropertySpec("forTeam", "Event", "Team", "forTeam", object_property=True),
    PropertySpec("time", "Event", str(XSD.string), "time"),
    PropertySpec("period", "Event", str(XSD.int), "period"),
    PropertySpec("secondsIntoGame", "Event", str(XSD.int), "secondsIntoGame"),
    PropertySpec("secondsLeftInPeriod", "Event", str(XSD.int), "secondsLeftInPeriod"),
    PropertySpec("eventNumber", "Event", str(XSD.int), "eventNumber"),
    PropertySpec("previousEvent", "Event", "Event", "previousEvent", object_property=True),
    PropertySpec("nextEvent", "Event", "Event", "nextEvent", object_property=True),
    PropertySpec("secondsSincePreviousEvent", "Event", str(XSD.int), "secondsSincePreviousEvent"),
    PropertySpec("secondsUntilNextEvent", "Event", str(XSD.int), "secondsUntilNextEvent"),
    PropertySpec("homeScore", "Event", str(XSD.int), "homeScore"),
    PropertySpec("awayScore", "Event", str(XSD.int), "awayScore"),
    PropertySpec("espnPlayId", "Event", str(XSD.string), "espnPlayId"),
    PropertySpec("espnSequenceNumber", "Event", str(XSD.string), "espnSequenceNumber"),
    PropertySpec("isScoringPlay", "Event", str(XSD.boolean), "isScoringPlay"),
    PropertySpec("isShootingPlay", "Event", str(XSD.boolean), "isShootingPlay"),
    PropertySpec("wallclockTime", "Event", str(XSD.dateTime), "wallclockTime"),
    # Team/player ids
    PropertySpec("espnTeamId", "Team", str(XSD.string), "espnTeamId"),
    PropertySpec("espnAthleteId", "Player", str(XSD.string), "espnAthleteId"),
    # Play type + coordinate
    PropertySpec("hasPlayType", "Play", "PlayType", "hasPlayType", object_property=True),
    PropertySpec("playTypeId", "PlayType", str(XSD.string), "playTypeId"),
    PropertySpec("playTypeText", "PlayType", str(XSD.string), "playTypeText"),
    PropertySpec("shortDescription", "Play", str(XSD.string), "shortDescription"),
    PropertySpec("hasCoordinate", "Play", "CourtCoordinate", "hasCoordinate", object_property=True),
    PropertySpec("coordinateX", "CourtCoordinate", str(XSD.int), "coordinateX"),
    PropertySpec("coordinateY", "CourtCoordinate", str(XSD.int), "coordinateY"),
    PropertySpec("pointsAttempted", "Play", str(XSD.int), "pointsAttempted"),
    PropertySpec("scoreValue", "Play", str(XSD.int), "scoreValue"),
    # Player involvement
    PropertySpec("involvedPlayer", "Play", "Player", "involvedPlayer", object_property=True),
    PropertySpec("shotBy", "Shot", "Player", "shotBy", object_property=True, subproperty_of="involvedPlayer"),
    PropertySpec(
        "shotAssistedBy", "Shot", "Player", "shotAssistedBy", object_property=True, subproperty_of="involvedPlayer"
    ),
    PropertySpec(
        "shotBlockedBy", "Block", "Player", "shotBlockedBy", object_property=True, subproperty_of="involvedPlayer"
    ),
    PropertySpec(
        "foulCommittedBy", "Foul", "Player", "foulCommittedBy", object_property=True, subproperty_of="involvedPlayer"
    ),
    PropertySpec("foulDrawnBy", "Foul", "Player", "foulDrawnBy", object_property=True, subproperty_of="involvedPlayer"),
    PropertySpec("reboundedBy", "Rebound", "Player", "reboundedBy", object_property=True, subproperty_of="involvedPlayer"),
    PropertySpec("turnedOverBy", "Turnover", "Player", "turnedOverBy", object_property=True, subproperty_of="involvedPlayer"),
    PropertySpec("stolenBy", "Turnover", "Player", "stolenBy", object_property=True, subproperty_of="involvedPlayer"),
    PropertySpec("playerEntering", "Enters", "Player", "playerEntering", object_property=True, subproperty_of="involvedPlayer"),
    PropertySpec("playerExiting", "Enters", "Player", "playerExiting", object_property=True, subproperty_of="involvedPlayer"),
    PropertySpec("playerEjected", "Ejection", "Player", "playerEjected", object_property=True, subproperty_of="involvedPlayer"),
    PropertySpec(
        "jumpBallHomePlayer", "JumpBall", "Player", "jumpBallHomePlayer", object_property=True, subproperty_of="involvedPlayer"
    ),
    PropertySpec(
        "jumpBallAwayPlayer", "JumpBall", "Player", "jumpBallAwayPlayer", object_property=True, subproperty_of="involvedPlayer"
    ),
    PropertySpec(
        "jumpBallGainedPossession",
        "JumpBall",
        "Player",
        "jumpBallGainedPossession",
        object_property=True,
        subproperty_of="involvedPlayer",
    ),
    # Play-specific data
    PropertySpec("shotMade", "Shot", str(XSD.boolean), "shotMade"),
    PropertySpec("shotType", "Shot", str(XSD.string), "shotType"),
    PropertySpec("shotPoints", "Shot", str(XSD.int), "shotPoints"),
    PropertySpec("turnoverType", "Turnover", str(XSD.string), "turnoverType"),
    PropertySpec("timeoutDuration", "Timeout", str(XSD.string), "timeoutDuration"),
    PropertySpec("isOfficial", "Timeout", str(XSD.boolean), "isOfficial"),
    PropertySpec("isShootingFoul", "Foul", str(XSD.boolean), "isShootingFoul"),
    PropertySpec("isLooseBallFoul", "Foul", str(XSD.boolean), "isLooseBallFoul"),
    PropertySpec("isCharge", "Foul", str(XSD.boolean), "isCharge"),
    PropertySpec("isPersonalBlockingFoul", "Foul", str(XSD.boolean), "isPersonalBlockingFoul"),
    PropertySpec("isOffensive", "Foul", str(XSD.boolean), "isOffensive"),
    PropertySpec("isThreeSecond", "TechnicalFoul", str(XSD.boolean), "isThreeSecond"),
    PropertySpec("isDelayOfGame", "TechnicalFoul", str(XSD.boolean), "isDelayOfGame"),
    PropertySpec("technicalFoulNumber", "TechnicalFoul", str(XSD.int), "technicalFoulNumber"),
    # Phase 2: win probability
    PropertySpec(
        "hasWinProbabilitySnapshot",
        "Game",
        "WinProbabilitySnapshot",
        "hasWinProbabilitySnapshot",
        object_property=True,
    ),
    PropertySpec(
        "snapshotForPlay",
        "WinProbabilitySnapshot",
        "Event",
        "snapshotForPlay",
        object_property=True,
    ),
    PropertySpec("homeWinProbability", "WinProbabilitySnapshot", str(XSD.decimal), "homeWinProbability"),
    PropertySpec("tieProbability", "WinProbabilitySnapshot", str(XSD.decimal), "tieProbability"),
    # Phase 2: venue + officials
    PropertySpec("venue", "Game", "Venue", "venue", object_property=True),
    PropertySpec("venueName", "Venue", str(XSD.string), "venueName"),
    PropertySpec("venueId", "Venue", str(XSD.string), "venueId"),
    PropertySpec("venueAddress", "Venue", str(XSD.string), "venueAddress"),
    PropertySpec("attendance", "Game", str(XSD.int), "attendance"),
    PropertySpec("hasOfficial", "Game", "Official", "hasOfficial", object_property=True),
    PropertySpec("officialName", "Official", str(XSD.string), "officialName"),
    PropertySpec("officialPosition", "Official", str(XSD.string), "officialPosition"),
]


def _class_iri(name: str) -> URIRef:
    return PBPRDF[name]


def _range_iri(range_name_or_iri: str) -> URIRef:
    if range_name_or_iri.startswith("http://"):
        return URIRef(range_name_or_iri)
    return _class_iri(range_name_or_iri)


def generate_ontology() -> Graph:
    graph = Graph()
    graph.bind("pbprdf", PBPRDF)
    graph.bind("owl", OWL)
    graph.bind("rdfs", RDFS)
    graph.bind("xsd", XSD)

    for class_name in CLASS_NAMES:
        class_iri = _class_iri(class_name)
        graph.add((class_iri, RDF.type, OWL.Class))
        graph.add((class_iri, RDFS.label, Literal(class_name)))

    graph.add((PBPRDF.Block, RDFS.subClassOf, PBPRDF.Shot))
    graph.add((PBPRDF.Play, RDFS.subClassOf, PBPRDF.Event))

    for spec in PROPERTY_SPECS:
        prop = PBPRDF[spec.name]
        graph.add((prop, RDF.type, OWL.ObjectProperty if spec.object_property else OWL.DatatypeProperty))
        graph.add((prop, RDFS.label, Literal(spec.label)))
        graph.add((prop, RDFS.domain, _class_iri(spec.domain)))
        graph.add((prop, RDFS.range, _range_iri(spec.range)))
        if spec.subproperty_of:
            graph.add((prop, RDFS.subPropertyOf, PBPRDF[spec.subproperty_of]))

    return graph

